demos/ota/aws_iot_ota_update_demo.c

```diff
@@ -48,242 +50,257 @@
+ * `EXIT_FAILURE` otherwise.
+ */
+static int _initializeOtaDemo( void )
+{
+    int status = EXIT_SUCCESS;
+    IotMqttError_t mqttInitStatus = IOT_MQTT_SUCCESS;
+
+    /* Initialize the MQTT library.*/
+    if( status == EXIT_SUCCESS )
     {
-        /* If the publish task is already connected to this network, set connected network flag to none,
-         * to trigger a reconnect.
-         */
-        if( ( xNetworkConnected == pdTRUE ) && ( xConnection.ulNetworkType == ulNetworkType ) )
+        mqttInitStatus = IotMqtt_Init();
+
+        if( mqttInitStatus != IOT_MQTT_SUCCESS )
         {
-            xNetworkConnected = pdFALSE;
+            /* Failed to initialize MQTT library.*/
+            status = EXIT_FAILURE;
         }
     }
-}
 
+    return status;
+}
 
-static IotNetworkError_t prvNetworkDisconnectCallback( void * pvContext )
+/**
+ * @brief Clean up libraries initialized for OTA demo.
+ */
+static void _cleanupOtaDemo( void )
 {
-    ( void ) pvContext;
-    xNetworkConnected = pdFALSE;
-
-    return IOT_NETWORK_SUCCESS;
+    /* Cleanup MQTT library.*/
+    IotMqtt_Cleanup();
 }
 
-static BaseType_t prxCreateNetworkConnection( void )
+/**
+ * @brief Delay before retrying network connection upto a maximum interval.
+ */
+static void _connectionRetryDelay( void )
 {
-    BaseType_t xRet = pdFALSE;
-    uint32_t ulRetriesLeft = otaDemoCONN_RETRY_LIMIT;
-    uint32_t ulRetryIntervalMS = otaDemoCONN_RETRY_INTERVAL_MS;
+    TickType_t intervalTicks = 0;
 
-    do
+    retryInterval *= 2;
+
+    if( retryInterval > OTA_DEMO_CONN_RETRY_MAX_INTERVAL_SECONDS )
     {
-        /* No networks are available, block for a physical network connection. */
-        if( ( AwsIotNetworkManager_GetConnectedNetworks() & otaDemoNETWORK_TYPES ) == AWSIOT_NETWORK_TYPE_NONE )
-        {
-            /* Block for a Network Connection. */
-            configPRINTF( ( "Waiting for a network connection.\r\n" ) );
-            ( void ) xSemaphoreTake( xNetworkAvailableLock, portMAX_DELAY );
-        }
+        retryInterval = OTA_DEMO_CONN_RETRY_MAX_INTERVAL_SECONDS;
+    }
 
-        #if defined( CONFIG_OTA_UPDATE_DEMO_ENABLED )
-            /* Connect to one of the network type.*/
-            xRet = xMqttDemoCreateNetworkConnection( &xConnection, otaDemoNETWORK_TYPES );
-        #endif
+    configPRINTF( ( "Retrying network connection in %d Secs ", retryInterval ) );
 
-        if( xRet == pdTRUE )
-        {
-            break;
-        }
-        else
-        {
-            /* Connection failed. Retry for a configured number of retries. */
-            if( ulRetriesLeft > 0 )
-            {
-                configPRINTF( ( "Network Connection failed, retry delay %lu ms, retries left %lu\r\n", ulRetryIntervalMS, ulRetriesLeft ) );
-                vTaskDelay( pdMS_TO_TICKS( ulRetryIntervalMS ) );
-            }
-        }
-    } while( ulRetriesLeft-- > 0 );
+    intervalTicks = pdMS_TO_TICKS( retryInterval * 1000 );
```

Comment:
Ideally we would add some jitter here.

Reply:
Adding jitter using rand so devices will initiate reconnection anywhere within the interval window.